ServiceEntry IP auto allocation - inconsistent use of hashing causes extra IP changes

### Bug Description

On 1.17.1 we noticed a sudden BlackHoleCluster issue across many namespaces

We have identified one issue as the likely root cause, with the new hash based IP auto allocation.

# Problem
## Walk through of current implementation
First, a hash is created based on namespace and hostname:

```go
	hashedServices := make([]*model.Service, maxIPs)
        // ...
			hash.Write([]byte(svc.Attributes.Namespace + "/" + svc.Hostname.String()))
                        hashedServices[firstHash] = svc
```

then, later, they are de-duplicated based on only the hostname:

```go
	for _, svc := range hashedServices {
                // ...
		n := svc.Hostname.String()
		if v, ok := hnMap[n]; ok {
			log.Debugf("Reuse IP for domain %s", n)
			setAutoAllocatedIPs(svc, v)
```
## Issue
This introduces a problem when a namespace with a low value of `hash(namespace + "/" + hostname)` is introduced or removed, as the IP will change across all namespaces.

## Issue shown as unit test:

```go
		{
			name: "order independent regardless of namespace - part 1",
			inServices: []*model.Service{
				{
					Hostname:       "a.example.com",
					Resolution:     model.DNSLB,
					DefaultAddress: "0.0.0.0",
					Attributes:     model.ServiceAttributes{Namespace: "a"},
				},
			},
			wantServices: []*model.Service{
				{
					Hostname:                 "a.example.com",
					Resolution:               model.DNSLB,
					DefaultAddress:           "0.0.0.0",
					AutoAllocatedIPv4Address: "240.240.163.38",
					AutoAllocatedIPv6Address: "2001:2::f0f0:a326",
				},
			},
		},
		{
			name: "order independent regardless of namespace - part 2",
			inServices: []*model.Service{
				{
					Hostname:       "a.example.com",
					Resolution:     model.DNSLB,
					DefaultAddress: "0.0.0.0",
					Attributes:     model.ServiceAttributes{Namespace: "a"},
				},
				{
					Hostname:       "a.example.com",
					Resolution:     model.DNSLB,
					DefaultAddress: "0.0.0.0",
					Attributes:     model.ServiceAttributes{Namespace: "b"},
				},
			},
			wantServices: []*model.Service{
				{
					Hostname:                 "a.example.com",
					Resolution:               model.DNSLB,
					DefaultAddress:           "0.0.0.0",
					AutoAllocatedIPv4Address: "240.240.163.38",
					AutoAllocatedIPv6Address: "2001:2::f0f0:a326",
				},
				{
					Hostname:                 "a.example.com",
					Resolution:               model.DNSLB,
					DefaultAddress:           "0.0.0.0",
					AutoAllocatedIPv4Address: "240.240.163.38",
					AutoAllocatedIPv6Address: "2001:2::f0f0:a326",
				},
			},
		},
```

output, we can see the IP changed:

```
=== RUN   Test_autoAllocateIP_conditions/order_independent_regardless_of_namespace_-_part_1
=== RUN   Test_autoAllocateIP_conditions/order_independent_regardless_of_namespace_-_part_2
2023-03-09T13:00:43.416996Z	info	serviceentry	Reuse IP for domain a.example.com
    controller_test.go:1795: autoAllocateIPs() AutoAllocatedIPv4Address = 240.240.114.198, want 240.240.163.38
    controller_test.go:1799: autoAllocateIPs() AutoAllocatedIPv4Address = 2001:2::f0f0:72c6, want 2001:2::f0f0:a326
    controller_test.go:1795: autoAllocateIPs() AutoAllocatedIPv4Address = 240.240.114.198, want 240.240.163.38
    controller_test.go:1799: autoAllocateIPs() AutoAllocatedIPv4Address = 2001:2::f0f0:72c6, want 2001:2::f0f0:a326
--- FAIL: Test_autoAllocateIP_conditions (0.00s)
    --- PASS: Test_autoAllocateIP_conditions/no_allocation_for_passthrough (0.00s)
    --- PASS: Test_autoAllocateIP_conditions/no_allocation_if_address_exists (0.00s)
    --- PASS: Test_autoAllocateIP_conditions/no_allocation_if_hostname_is_wildcard (0.00s)
    --- PASS: Test_autoAllocateIP_conditions/allocate_IP_for_clientside_lb (0.00s)
    --- PASS: Test_autoAllocateIP_conditions/allocate_IP_for_dns_lb (0.00s)
    --- PASS: Test_autoAllocateIP_conditions/collision (0.00s)
    --- PASS: Test_autoAllocateIP_conditions/order_independent_regardless_of_namespace_-_part_1 (0.00s)
    --- FAIL: Test_autoAllocateIP_conditions/order_independent_regardless_of_namespace_-_part_2 (0.00s)

```

## Fix
I believe simply removing the namespace from the first hash makes it stable again.

IPs in unit tests needs to be updated.

This will cause *all* IPs to change during the next upgrade.

### Version

```prose
$ istioctl version
client version: 1.16.1
control plane version: 1.17.1
data plane version: 1.16.1 (167 proxies), 1.17.1 (198 proxies)
```


### Additional Information

_No response_

--- a/pilot/pkg/serviceregistry/serviceentry/controller.go
+++ b/pilot/pkg/serviceregistry/serviceentry/controller.go
@@ -890,7 +890,7 @@ func autoAllocateIPs(services []*model.Service) []*model.Service {
 		// 3. the hostname is not a wildcard
 		if svc.DefaultAddress == constants.UnspecifiedIP && !svc.Hostname.IsWildCarded() &&
 			svc.Resolution != model.Passthrough {
-			hash.Write([]byte(svc.Attributes.Namespace + "/" + svc.Hostname.String()))
+			hash.Write([]byte(makeServiceKey(svc)))
 			// First hash is calculated by
 			s := hash.Sum32()
 			firstHash := s % uint32(maxIPs)
@@ -933,7 +933,7 @@ func autoAllocateIPs(services []*model.Service) []*model.Service {
 			}
 			continue
 		}
-		n := svc.Hostname.String()
+		n := makeServiceKey(svc)
 		if v, ok := hnMap[n]; ok {
 			log.Debugf("Reuse IP for domain %s", n)
 			setAutoAllocatedIPs(svc, v)
@@ -953,6 +953,10 @@ func autoAllocateIPs(services []*model.Service) []*model.Service {
 		}
 	}
 	return services
+}
+
+func makeServiceKey(svc *model.Service) string {
+	return svc.Attributes.Namespace + "/" + svc.Hostname.String()
 }
 
 func setAutoAllocatedIPs(svc *model.Service, octets octetPair) {

--- a/pilot/pkg/serviceregistry/serviceentry/controller_test.go
+++ b/pilot/pkg/serviceregistry/serviceentry/controller_test.go
@@ -1687,6 +1687,37 @@ func Test_autoAllocateIP_conditions(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "collision",
+			inServices: []*model.Service{
+				{
+					Hostname:       "a17061.example.com",
+					Resolution:     model.DNSLB,
+					DefaultAddress: "0.0.0.0",
+				},
+				{
+					Hostname:       "a44155.example.com",
+					Resolution:     model.DNSLB,
+					DefaultAddress: "0.0.0.0",
+				},
+			},
+			wantServices: []*model.Service{
+				{
+					Hostname:                 "a17061.example.com",
+					Resolution:               model.DNSLB,
+					DefaultAddress:           "0.0.0.0",
+					AutoAllocatedIPv4Address: "240.240.0.1",
+					AutoAllocatedIPv6Address: "2001:2::f0f0:1",
+				},
+				{
+					Hostname:                 "a44155.example.com",
+					Resolution:               model.DNSLB,
+					DefaultAddress:           "0.0.0.0",
+					AutoAllocatedIPv4Address: "240.240.75.79",
+					AutoAllocatedIPv6Address: "2001:2::f0f0:4b4f",
+				},
+			},
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

--- a/pilot/pkg/serviceregistry/serviceentry/controller_test.go
+++ b/pilot/pkg/serviceregistry/serviceentry/controller_test.go
@@ -1690,14 +1690,16 @@ func Test_autoAllocateIP_conditions(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			got := autoAllocateIPs(tt.inServices)
-			if got[0].AutoAllocatedIPv4Address != tt.wantServices[0].AutoAllocatedIPv4Address {
-				t.Errorf("autoAllocateIPs() AutoAllocatedIPv4Address = %v, want %v",
-					got[0].AutoAllocatedIPv4Address, tt.wantServices[0].AutoAllocatedIPv4Address)
-			}
-			if got[0].AutoAllocatedIPv6Address != tt.wantServices[0].AutoAllocatedIPv6Address {
-				t.Errorf("autoAllocateIPs() AutoAllocatedIPv4Address = %v, want %v",
-					got[0].AutoAllocatedIPv6Address, tt.wantServices[0].AutoAllocatedIPv6Address)
+			gotServices := autoAllocateIPs(tt.inServices)
+			for i, got := range gotServices {
+				if got.AutoAllocatedIPv4Address != tt.wantServices[i].AutoAllocatedIPv4Address {
+					t.Errorf("autoAllocateIPs() AutoAllocatedIPv4Address = %v, want %v",
+						got.AutoAllocatedIPv4Address, tt.wantServices[i].AutoAllocatedIPv4Address)
+				}
+				if got.AutoAllocatedIPv6Address != tt.wantServices[i].AutoAllocatedIPv6Address {
+					t.Errorf("autoAllocateIPs() AutoAllocatedIPv4Address = %v, want %v",
+						got.AutoAllocatedIPv6Address, tt.wantServices[i].AutoAllocatedIPv6Address)
+				}
 			}
 		})
 	}

--- a/pilot/pkg/serviceregistry/serviceentry/controller_test.go
+++ b/pilot/pkg/serviceregistry/serviceentry/controller_test.go
@@ -1696,6 +1696,8 @@ func Test_autoAllocateIP_conditions(t *testing.T) {
 					DefaultAddress: "0.0.0.0",
 				},
 				{
+					// hashes to the same value as the hostname above,
+					// a new collision needs to be found if the hash algorithm changes
 					Hostname:       "a44155.example.com",
 					Resolution:     model.DNSLB,
 					DefaultAddress: "0.0.0.0",
@@ -1715,6 +1717,59 @@ func Test_autoAllocateIP_conditions(t *testing.T) {
 					DefaultAddress:           "0.0.0.0",
 					AutoAllocatedIPv4Address: "240.240.75.79",
 					AutoAllocatedIPv6Address: "2001:2::f0f0:4b4f",
+				},
+			},
+		},
+		{
+			name: "stable IP - baseline test",
+			inServices: []*model.Service{
+				{
+					Hostname:       "a.example.com",
+					Resolution:     model.DNSLB,
+					DefaultAddress: "0.0.0.0",
+					Attributes:     model.ServiceAttributes{Namespace: "a"},
+				},
+			},
+			wantServices: []*model.Service{
+				{
+					Hostname:                 "a.example.com",
+					Resolution:               model.DNSLB,
+					DefaultAddress:           "0.0.0.0",
+					AutoAllocatedIPv4Address: "240.240.163.38",
+					AutoAllocatedIPv6Address: "2001:2::f0f0:a326",
+				},
+			},
+		},
+		{
+			name: "stable IP - not affected by other namespace",
+			inServices: []*model.Service{
+				{
+					Hostname:       "a.example.com",
+					Resolution:     model.DNSLB,
+					DefaultAddress: "0.0.0.0",
+					Attributes:     model.ServiceAttributes{Namespace: "a"},
+				},
+				{
+					Hostname:       "a.example.com",
+					Resolution:     model.DNSLB,
+					DefaultAddress: "0.0.0.0",
+					Attributes:     model.ServiceAttributes{Namespace: "b"},
+				},
+			},
+			wantServices: []*model.Service{
+				{
+					Hostname:                 "a.example.com",
+					Resolution:               model.DNSLB,
+					DefaultAddress:           "0.0.0.0",
+					AutoAllocatedIPv4Address: "240.240.163.38",
+					AutoAllocatedIPv6Address: "2001:2::f0f0:a326",
+				},
+				{
+					Hostname:                 "a.example.com",
+					Resolution:               model.DNSLB,
+					DefaultAddress:           "0.0.0.0",
+					AutoAllocatedIPv4Address: "240.240.114.198",
+					AutoAllocatedIPv6Address: "2001:2::f0f0:72c6",
 				},
 			},
 		},

--- a/releasenotes/notes/43858.yaml
+++ b/releasenotes/notes/43858.yaml
@@ -1,0 +1,10 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+
+issue:
+  - 43858
+
+releaseNotes:
+- |
+  **Fixed** an issue where auto allocated service entry IPs change on host reuse

